services/web/server/src/simcore_service_webserver/tags_handlers.py

```diff
@@ -1,84 +1,86 @@
-import sqlalchemy as sa
+import functools
+
 from aiohttp import web
 from servicelib.aiohttp.application_keys import APP_DB_ENGINE_KEY
-from sqlalchemy import and_
+from servicelib.aiohttp.typing_extension import Handler
+from servicelib.mimetype_constants import MIMETYPE_APPLICATION_JSON
+from simcore_postgres_database.utils_tags import (
+    TagNotFoundError,
+    TagOperationNotAllowed,
+    TagsRepo,
+    ValidationError,
+)
 
-from .db_models import tags
 from .login.decorators import RQT_USERID_KEY, login_required
 from .security_api import check_permission
 
 
+def _handle_tags_exceptions(handler: Handler):
+    @functools.wraps(handler)
+    async def wrapper(request: web.Request) -> web.Response:
+        try:
+            return await handler(request)
+
+        except ValidationError as exc:
+            raise web.HTTPBadRequest(reason=f"{exc}") from exc
+
+        except TagNotFoundError as exc:
+            raise web.HTTPNotFound(reason=f"{exc}") from exc
+
+        except TagOperationNotAllowed as exc:
+            raise web.HTTPUnauthorized(reason=f"{exc}") from exc
+
+    return wrapper
+
+
 @login_required
+@_handle_tags_exceptions
```

Comment:
Question: I'm wondering if these can be setup as middlewares. Also if there is an advantage/disadvantage.

Reply:
It is possible to use middleware but IMO it is less efficient. 
These handlers are only for tags entrypoints while middlewares by default apply to all entrypoints. Instead of having an "if" statement in a middleware I think it is better to simply add the decorator to the corresponding handler at "import time"